(Mathematical-Functions-Mathematical-Functions)=
# Mathematical Functions

(Mathematical-Functions-Sorting-and-Searching-Data)=
## Sorting and Searching Data

(Mathematical-Functions-Sorting-Numeric-Data)=
### Sorting Numeric Data

(Mathematical-Functions-Sorting-Numeric-Vectors)=
#### Sorting Numeric Vectors

Two primitive functions are provided to sort data:

 - _Grade up_ `⍋`, typed with <kbd>APL</kbd>+<kbd>Shift</kbd>+<kbd>4</kbd>, returns the set of indices required to sort the array in ascending order.
 - _Grade down_ `⍒`, typed with <kbd>APL</kbd>+<kbd>Shift</kbd>+<kbd>3</kbd>, returns the set of indices required to sort the array in descending order.

Here is an example:

In [1]:
vec ← 40 60 20 110 70 60 40 90 60 80
⍋vec

3 1 7 2 6 9 5 10 8 4

Notice that the function _grade up_ `⍋` does not actually return a sorted (i.e., re-ordered) array.
Instead, we obtain a set of indices which tell us that:

 - the smallest item is at index `3`;
 - then comes the item at index `1`;
 - then the item at index `7`;
 - then the item at index `2`;
 - ...
 - and the largest item is at index `4`.

To obtain the vector `vec` sorted in ascending order, we must index it by these values, as shown below:

In [2]:
vec[⍋vec]

20 40 40 60 60 60 70 80 90 110

As you might imagine, _grade down_ sorts the vector in descending order:

In [3]:
vec[⍒vec]

110 90 80 70 60 60 60 40 40 20

When several items of the argument are equal, the relative positions of the items that are equal do not change.
For this reason, `⍒vec` is equal to `⌽⍋vec` only if the values in `vec` are all different.
This is not the case for our vector:

In [4]:
⍋vec

3 1 7 2 6 9 5 10 8 4

In [5]:
⌽⍋vec

4 8 10 5 9 6 2 7 1 3

In [6]:
⍒vec

4 8 10 5 2 6 9 1 7 3

Let us compare `⌽⍋vec` with `⍒vec`:

```APL
4 8 10 5 9 6 2 7 1 3  ⍝ ⌽⍋vec
4 8 10 5 2 6 9 1 7 3  ⍝  ⍒vec
         ↑ ↑ ↑
```

Notice how the numbers `2 6 9` are in opposite directions in the two results.
That is because those indices refer to the three times the number 60 shows up in `vec`.

Both when using _grade up_ and _grade down_, the indices show up in the order `2 6 9` because both grades preserve the relative positions of the items that are the same.
Thus, when we _reverse_ the result of _grade up_, those three numbers show up as `9 6 2`.

You may rightfully wonder why the functions _grade up_ and _grade down_ return indices instead of sorted data.
There are, of course, good reasons for it, as we will show here.

Firstly, the availability of the intermediate result of the sorting process opens up possibilities for interesting and useful manipulations.
Here is one example: given a vector `v` of unique values, the expression `⍋⍋v` (or `⍋⍒v`) indicates which position the items of `v` would occupy if they were sorted in increasing (or decreasing) order.

In [7]:
class ← 153 432 317 609 411 227 186 350
⍋⍒class

8 2 5 1 3 6 7 4

In [8]:
class,[.5]⍋⍒class

153 432 317 609 411 227 186 350
 8 2 5 1 3 6 7 4

The result above shows that `153` would be sent to position `8` if `class` were ordered in decreasing order, the `432` would be in position `2`, and so on.
Here is the verification:

In [9]:
⎕← classSorted ← class[⍒class]

609 432 411 350 317 227 186 153

In [10]:
classSorted[8]≡153

1

In [11]:
classSorted[2]≡432

1

Secondly, it makes it much easier to sort complementary arrays (e.g., arrays that represent columns in a database table) in the same order.
Suppose, for example, that we would like to sort our familiar lists of prices and quantities in ascending order of quantity:

In [12]:
price ← 5.2 11.5 3.6 4 8.45
qty ← 2 1 3 6 2
idx ← ⍋qty

In [13]:
⎕← price ← price[idx]

11.5 5.2 8.45 3.6 4

In [14]:
⎕← qty ← qty[idx]

1 2 2 3 6

(Mathematical-Functions-Sorting-Numeric-Matrices)=
#### Sorting Numeric Matrices

The functions _grade up_ and _grade down_ can be used to sort the rows of a matrix.

In this case, they sort the rows by comparing the first element of each row; then, for rows that have the same first element, they compare second elements; and so on and so forth.
Both functions return a vector of row indices, which can be used to sort the matrix (do not forget the semicolon when indexing).

In [15]:
⎕← matrix ← 7 3⍴2 40 8 5 55 2 2 33 9 7 20 2 5 55 1 5 52 9 2 40 9

2 40 8
5 55 2
2 33 9
7 20 2
5 55 1
5 52 9
2 40 9

In [16]:
matrix[⍋matrix;]

2 33 9
2 40 8
2 40 9
5 52 9
5 55 1
5 55 2
7 20 2

In the result above, you can see that the first column is sorted in ascending order.
Then, when the first column has the same element, the second column is sorted in ascending order.
For example, there are three rows in `matrix` that start with a `2`.
Those three rows are the first three rows of the sorted matrix and, among those three rows, the second column is sorted:

In [17]:
3 ¯2↑matrix[⍋matrix;]

33 9
40 8
40 9

Finally, when the first and second columns are the same, we order by the third column.
In the example above, we can see that two rows started with `2 40`, and those two rows were then sorted according to their third column.

(Mathematical-Functions-Sorting-Numeric-High-rank-Arrays)=
#### Sorting Numeric High-rank Arrays

The functions _grade up_ and _grade down_ can be used to sort the major cells of numeric arrays of arbitrary rank, not just vectors and matrices.

To sort the rows of a matrix, we compared the rows item by item.
To sort the planes of a cuboid, we will compare the planes row by row.
To sort the cuboids of a 4D array, we will compare the cuboids plane by plane.
And this is the reasoning that you should follow to sort the major cells of any arbitrary array.

Consider the following cuboid:

In [18]:
⎕← cuboid ← 5 2 2⍴2 1 2 2 1 2 2 2 1 2 1 1 2 1 2 1 2 2 2 2

2 1
2 2

1 2
2 2

1 2
1 1

2 1
2 1

2 2
2 2

Looking at all of its planes, we see two matrices whose first row is `1 2`:

In [19]:
cuboid[2 3;;]

1 2
2 2

1 2
1 1

Those two matrices have the smallest first row, so those two matrices will be the first two matrices in the sorted cuboid:

In [20]:
2↑cuboid[⍋cuboid;;]

1 2
1 1

1 2
2 2

However, they show up in the reverse order in the sorted cuboid because the two matrices have the same first row, thus we need to compare the second row of each matrix to determine which matrix comes first.

This is the type of reasoning that must be followed to sort the whole cuboid:

In [21]:
cuboid[⍋cuboid;;]

1 2
1 1

1 2
2 2

2 1
2 1

2 1
2 2

2 2
2 2

(Mathematical-Functions-Sorting-Characters)=
### Sorting Characters

(Mathematical-Functions-Using-the-Default-Alphabet)=
#### Using the Default Alphabet

When applied to characters, the monadic forms of _grade up_ and _grade down_ refer to an implicit alphabetic order.
In all modern versions of Dyalog (the _Unicode Editions_), it is the numerical order of the corresponding Unicode code points.
However, in the _Classic Editions_ of Dyalog, it is given by a specific _system variable_ known as the _atomic vector_ `⎕AV`, described in [the chapter about system interfaces](./System-Interfaces.ipynb).

For example, all modern versions of Dyalog give this result:

In [22]:
text ← 'Grade Up also works on Characters'
text[⍋text]

CGUaaaacdeehklnoooprrrrssstw

In _Classic Editions_ of Dyalog, the result would have been `'     aaaacdeehjlnoooprrrrssstwCGU'`.
That is because the vector `⎕AV` has the lower case characters before the upper case ones, which means all the lower case letters will be sorted before all the upper case letters.
However, in the _Unicode Editions_, text is sorted according to Unicode code points, and in the Unicode standard, the upper case letters come before the lower case letters.
This is why we obtained two different results.

For this reason, sorting characters using the default alphabet should be reserved for text that contains only lower case or only upper case letters, or matrices where upper and lower case letters appear in the same columns, as in the following example:

In [23]:
⎕← towns ← ↑'Canberra' 'Paris' 'Washington' 'Moscow' 'Martigues' 'Mexico'

Canberra 
Paris 
Washington
Moscow 
Martigues 
Mexico

In [24]:
towns[⍋towns;]

Canberra 
Martigues 
Mexico 
Moscow 
Paris 
Washington

The names are sorted correctly because all letters in each column are of the same case (all are upper case, or all are lower case).
However, if we mix upper and lower case letters, we may not get the result we expected.

For example, here we redefine the matrix `towns` but we are not careful, and thus do not capitalise properly all of the names of the cities:

In [25]:
⎕← towns ← ↑'canberra' 'paris' 'Washington' 'Moscow' 'Martigues' 'Mexico'

canberra 
paris 
Washington
Moscow 
Martigues 
Mexico

Because capitalisation was not uniform, sorting will differ wildly from what we obtained before:

In [26]:
towns[⍋towns;]

Martigues 
Mexico 
Moscow 
Washington
canberra 
paris

In the result above, `'Washington'` (which was supposed to be the last row) came before `'canberra'` (which was supposed to be the first row) because all the upper case letters come before the lower case letters in the implicit alphabet that defines the order of characters.

(Mathematical-Functions-Using-an-Explicit-Alphabet)=
#### Using an Explicit Alphabet

To avoid this kind of problem, it is advisable to use the dyadic versions of the sorting primitives, which take an explicit alphabet as their left argument (notice it contains a blank in the end).
Let us try this:

In [27]:
alph ← 'aAbBcCdDeEfFgGhHiIjJkKlLmMnNoOpPqQrRsStTuUvVwWxXyYzZ '
towns[alph⍋towns;]

canberra 
Martigues 
Mexico 
Moscow 
paris 
Washington

Alas, our satisfaction will be short lived!
Imagine that `'Mexico'` now becomes `'mexico'`:

In [28]:
towns[6;1] ← 'm'
⎕← towns

canberra 
paris 
Washington
Moscow 
Martigues 
mexico

In our alphabet, a lower case "m" comes before the upper case "M", so we would obtain:

In [29]:
towns[alph⍋towns;]

canberra 
mexico 
Martigues 
Moscow 
paris 
Washington

There is fortunately a solution to this annoying problem.
Instead of a vector, let us organise our alphabet into a matrix, as shown here:

In [30]:
⎕← alph ← ⍉27 2⍴alph,' '

abcdefghijklmnopqrstuvwxyz 
ABCDEFGHIJKLMNOPQRSTUVWXYZ

Although you cannot see it, note that the last column contains a blank in both rows:

In [31]:
alph,'#'

abcdefghijklmnopqrstuvwxyz #
ABCDEFGHIJKLMNOPQRSTUVWXYZ #

Now, "m" and "M" have the same priority because they are placed in the same columns of the alphabet.
When sorting a matrix, the lower case letters and the upper case letters will now be sorted identically:

In [32]:
towns[alph⍋towns;]

canberra 
Martigues 
mexico 
Moscow 
paris 
Washington

Note that if we include the same town twice, but with different capitalisation, the lower case version will come first:

In [33]:
towns ⍪← 'washington'
⎕← towns

canberra 
paris 
Washington
Moscow 
Martigues 
mexico 
washington

In [34]:
towns[alph⍋towns;]

canberra 
Martigues 
mexico 
Moscow 
paris 
washington
Washington

(Mathematical-Functions-Grading-and-Permutations)=
### Grading and Permutations

Suppose the function _grade up_ or _grade down_ is applied to a vector of length `n`.
The result of the function is going to be what mathematicians call a _permutation_.

A permutation (of size `n`) is a vector of length `n` that contains all integers from `1` to `n` in _any_ order.
Intuitively speaking, a permutation is the vector `⍳n` shuffled.

As we have seen above, permutation vectors can be used to index into other vectors (or arbitrary arrays) and change the orders of their contents.
That is one of the primary use cases of permutation vectors and that is, essentially, the intended use case of the permutation vectors returned by the functions _grade up_ and _grade down_:

In [35]:
word ← 'MISSISSIPPI'

In [36]:
⎕← permutationVector ← ⍋word

2 5 8 11 1 9 10 3 4 6 7

As we can see, the result is a permutation vector of size 11 because it contains all the integers from 1 to 11.
Then, we can use that permutation vector to reorder other vectors of length 11.
For example, we can use the permutation vector to sort the original character vector:

In [37]:
word[permutationVector]

IIIIMPPSSSS

When a vector (or another array) is indexed by a permutation vector, we know that:

 - the result of the indexing will have as many elements as the array being indexed;
 - no element of the array being indexed will be repeated; and
 - all elements of the array being indexed will be in the final result.

These characteristics of the result of the indexing operation all follow from the fact that the indexing vector is a permutation vector.

Finally, if you look at permutation vectors as vectors that allow you to change the order of things, sometimes you might want to change the order **back**.
If `p` is a permutation vector, then `⍋p` is the inverse permutation vector.
So, if you use the permutation vector `p` to reorder something, you can use the permutation vector `⍋p` to undo that reordering.

In other words, if `vp ← v[p]`, then `v ≡ vp[⍋p]`.

Here is an example:

In [38]:
wordP ← word[permutationVector]  ⍝ reorder the word
word ≡ wordP[⍋permutationVector]  ⍝ go back

1

(Mathematical-Functions-Finding-Values)=
### Finding Values

The function _find_ `⍷`, which you can type with <kbd>APL</kbd>+<kbd>Shift</kbd>+<kbd>E</kbd>, is a primitive function that allows you to search for an array `needle` in an array `haystack`.
The result is a Boolean array with the same shape as `haystack`, with a 1 at the starting point of each occurrence of `needle` in `haystack`.
Here is an example:

In [39]:
⎕← where ← 'at' ⍷ 'Congratulations'

0 0 0 0 0 1 0 0 0 1 0 0 0 0 0

In [40]:
'Congratulations',[.5]where

C o n g r a t u l a t i o n s
0 0 0 0 0 1 0 0 0 1 0 0 0 0 0

The function _find_ can also be applied to numeric arrays.
Here, we search for a vector of 3 numbers in a longer vector:

In [41]:
2 5 1 ⍷ 4 8 2 5 1 6 4 2 5 3 5 1 2 2 5 1 7

0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0

The rank of `haystack` can be higher than the rank of `needle`.
For example, we can search for a vector in a matrix:

In [42]:
'as' ⍷ towns

0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 1 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 1 0 0 0 0 0 0 0 0

The opposite is also permitted (i.e., the left argument having a higher rank than the right argument), but would not be very useful.
The result would be only zeroes.

(Mathematical-Functions-Encode-and-Decode)=
## Encode and Decode

APL offers two primitives, _encode_ `⊤` and _decode_ `⊥` – typed with <kbd>APL</kbd>+<kbd>n</kbd> and <kbd>APL</kbd>+<kbd>b</kbd>, respectively –, to convert numeric values from their decimal form to a representation in any other number system, and back again.

Because we may not be very familiar with this kind of calculation, it may seem that only mad mathematicians should invest their time studying such conversions.
In fact, these functions are used rather frequently to solve common problems.
But before studying them, we need to present some basic notions.

(Mathematical-Functions-Some-Words-of-Theory)=
### Some Words of Theory

(Mathematical-Functions-Familiar-But-Not-Decimal)=
#### Familiar, But Not Decimal

8839 is a simple number, represented in our good old decimal system.
But if 8839 represents a number of seconds, we could just express it as 2 hours, 27 minutes, and 19 seconds.
`2 27 19` is the representation of `8839` in a non-uniform number system based on 24 hour days, each divided into 60 minutes, each divided into 60 seconds.

The second representation is more familiar to us, but is **not** a decimal representation: the value has been expressed in a complex _base_ (or _radix_); we shall say that it is **coded**, even though it is familiar.

Converting `8839` into `2 27 19` is called _encoding_ because the result is not decimal.
Converting `2 27 19` into `8839` is called _decoding_ because the result is decimal.
We shall say that `24 60 60` is the _base_ of the number system of `2 27 19`.

(Mathematical-Functions-Three-Important-Remarks)=
#### Three Important Remarks

 - In this case, encoding a scalar (`8839`) produces a vector (`2 27 19`).
 In general, the representation of a decimal scalar in a non-decimal base cannot be expected to be a single number.
 It will always be an array of the same shape as the left argument to the function _encode_, and in all but very special cases this left argument will be a vector or a higher-rank array.

For example, a binary value cannot be written as `101011` (this is a decimal number); it must be written as a vector of binary digits: `1 0 1 0 1 1`.

 - The items of an encoded value can be greater than 9.
 In our examples, we had items equal to 27 and 19.
 But they are always smaller than the corresponding item of the _base_.
 Would you say that you spent 2 hours and 87 minutes to do something?
 Probably not, because 87 is greater than 60.
 You would say 3 hours and 27 minutes.

 - No mater whether days were made of only 18 hours, 2 hours, 27 minutes, and 19 seconds would always represent 8839 seconds.

This leads to the following rule: the first item of the base vector is never taken into account when _decoding_, but it is always used for _encoding_.

(Mathematical-Functions-Base-and-Weights)=
#### Base and Weights

Given the _base_ vector `24 60 60` and a value `2 27 19`, one can _decode_ it (obtain its decimal representation) by any of the three formulas that follow.
(Note that `3600=60×60`.)

In [43]:
(2×3600) + (27×60) + 19

8839

In [44]:
+/ 3600 60 1 × 2 27 19

8839

In [45]:
3600 60 1 +.× 2 27 19

8839

The last formula clearly shows that decoding a set of values is nothing else than an _inner product_.
This is important because it means that the same shape compatibility rules will apply.

The values `3600 60 1` are the _weights_ representing how many seconds are in an hour, in a minute, and in a second.
They can be obtained from the _base_ vector as follows:

In [46]:
base ← 24 60 60
⎕← weights ← ⌽ 1,×\ ⌽ 1↓base

3600 60 1

This formula confirms the remark we made earlier: the first item of the _base_ vector is not used when _decoding_ a value.
However, it is needed in order to do the reverse operation (_encoding_) using the same _base_ vector.

Once the weights are calculated, we can define the relationship between _decode_ and _inner product_:

```{admonition} Remark 
:class: tip
`base ⊥ values` is equivalent to `weights +.× values`.
```


(Mathematical-Functions-Using-Decode--Encode)=
### Using Decode & Encode

(Mathematical-Functions-Decode)=
#### Decode

_Decode_ is represented by `⊥`.
It accepts the _base_ vector directly as its left argument, so that you do not have to calculate the weights:

In [47]:
base ⊥ 2 27 19

8839

Because the first item of the base vector is not used when decoding, we could have obtained the same result in this way:

In [48]:
base ← 0 60 60
base ⊥ 2 27 19

8839

As another example application of the function _decode_, suppose eggs are packaged in boxes, each containing 6 packs of 6 eggs.

If we have 2 full boxes, plus 5 packs, plus 3 eggs, we can calculate the total number of eggs using any of the following expressions, each giving the same result:

In [49]:
(2×36) + (5×6) + 3

105

In [50]:
+/ 36 6 1 × 2 5 3

105

In [51]:
36 6 1 +.× 2 5 3

105

In [52]:
6 6 6 ⊥ 2 5 3

105

The first item of the base is not relevant when decoding, so we could have used any other element:

In [53]:
999 6 6 ⊥ 2 5 3

105

In [54]:
0 6 6 ⊥ 2 5 3

105

However, in this very special case, our base is uniform (`6 6 6`), which lets us write the last expression in a simpler way.
As usual, the scalar value is reused as appropriate:

In [55]:
6 ⊥ 2 5 3

105

This says that `2 5 3` is the base 6 representation of 105.

(Mathematical-Functions-Shape-Compatibility)=
#### Shape Compatibility

We said that _decode_ is nothing but a plain _inner product_, so the same shape compatibility rules must be satisfied.

Imagine that we have to convert two durations given in hours, minutes, and seconds, into just seconds.
The first duration is 2 hours, 27 minutes, and 19 seconds, and the second one is 5 hours, 3 minutes, and 48 seconds.
When we put those durations into a single variable, it is natural to express the data as a matrix, as shown here:

In [56]:
⎕← hms ← 2 3⍴2 27 19 5 3 48

2 27 19
5 3 48

But we cannot combine a 3-item vector (`base ← 24 60 60`) with a 2-row matrix (`hms`).
`base ⊥ hms` would cause a `LENGTH ERROR`.
We must transpose `hms` in order to make the lengths of the arguments compatible:

In [57]:
base ⊥ ⍉hms

8839 18228

The length of `base` is equal to the length of the first dimension of `⍉hms`.
The same rule applies to _inner product_.

(Mathematical-Functions-Encode)=
#### Encode

As an example of encoding a decimal number, we can encode 105 into base 6:

In [58]:
6 6 6 ⊤ 105

2 5 3

Please note that specifying a scalar as the left argument in the expression above does not give the same result:

In [59]:
6 ⊤ 105

3

The reason is that it is not really possible for APL to "reuse the scalar as appropriate" here, because what does "appropriate" mean in this case?
The left argument to _encode_ defines the number of digits in the new number system, so if we want or need three digits, we must specify three `6`s.
(In {numref}`Operators-The-Result-of-an-Inverse-Function` you find an example showing a clever way to have APL itself figure out the number of digits needed to properly encode a number.)

We can convert a number of seconds into hours, minutes, and seconds, like this:

In [60]:
24 60 60 ⊤ 23456

6 30 56

However, when converting 3 values, the results must be read carefully:

In [61]:
24 60 60 ⊤ 8839 23456 18228

2 6 5
27 30 3
19 56 48

Do not read these results horizontally: 8839 seconds are **not** equal to 2 hours, 5 minutes, and 2 seconds!
You must read the result vertically, and you will recognise the results we got earlier: `2 27 19`, `6 30 56`, and `5 3 48`.

(Mathematical-Functions-Limited-Encoding)=
#### Limited Encoding

The shape of the result of `bases ⊤ values` is equal to `bases ,⍥⍴ values`.

No specific rule is imposed on the arguments' shapes, but if the last dimension of the base is too small, APL proceeds to a limited encoding, as shown below:

In [62]:
24 60 60 ⊤ 8839  ⍝ full conversion

2 27 19

In [63]:
60 60 ⊤ 8839  ⍝ truncated result

27 19

In [64]:
60 ⊤ 8839  ⍝ ditto

19

The last two results are truncated to the length of the specified base vector, but nothing indicates that they have been truncated.
To avoid potential misinterpretation, it is common to use a leading zero as the first item of the base:

In [65]:
0 24 60 60 ⊤ 123456

1 10 17 36

In [66]:
0 60 60 ⊤ 123456

34 17 36

In [67]:
0 60 ⊤ 123456

2057 36

In [68]:
0 ⊤ 123456

123456

The first conversion states that 123456 seconds represent 1 day, 10 hours, 17 minutes, and 36 seconds.
In the second conversion, limited to hours, 1 day plus 10 hours gave 34 hours.
In the third conversion, limited to minutes, shows that 123456 seconds is 2057 minutes and 36 seconds.

(Mathematical-Functions-Using-Several-Simultaneous-Bases)=
#### Using Several Simultaneous Bases

If one needs to encode or decode several values in several different bases, _base_ will no longer be a vector, but a matrix.
However, this is a bit more complex and will be studied in {numref}`Mathematical-Functions-the-Specialists-Section`.

(Mathematical-Functions-Applications-of-Encoding-and-Decoding)=
### Applications of Encoding and Decoding

(Mathematical-Functions-Condense-or-Expand-Values)=
#### Condense or Expand Values

It is sometimes convenient to condense several values into a single one.
In general, this does not save much memory space, but it may be more convenient to manipulate a single value rather than several.
This can be achieved by decoding the values into a decimal number.

Say, for example, that you have a list of 5 rarely used settings that you need to save in a relational database.
Instead of creating 5 columns in the database table to hold the settings, you could _decode_ the 5 values into a decimal integer and save it in a single database column.

Often, it is convenient to select a base made of powers of 10, corresponding to the maximum number of digits of the given values, for example:

In [69]:
100 1000 10 100 ⊥ 35 681 7 24

35681724

The single value `35681724` contains the same information as the original vector, and because all base values used are powers of 10, it is fairly easy to recognise the original numbers.

The conversion base was built like this:

 - 100 for 35 which has 2 digits;
 - 1000 for 681 which has 3 digits;
 - 10 for 7 which has only 1 digit; and
 - 100 for 24 which has 2 digits.

Of course, the base vector must be built according to the largest values that can appear in each of the items, not just from an arbitrary number as in our small example.

The reverse transformation may be done by encoding the value using the same base:

In [70]:
100 1000 10 100 ⊤ 35681724

35 681 7 24

A similar technique may be used to separate the integer and decimal parts of positive numbers:

In [71]:
0 1⊤127.83 619.26 423.44 19.962

127 619 423 19
 0.83 0.26 0.44 0.962

(Mathematical-Functions-Calculating-Polynomials)=
#### Calculating Polynomials

Let us recall the example about packing eggs:

In [72]:
6⊥2 5 3

105

We can say that we used _decode_ to calculate `(2×6*2) + (5×6) + 3`.
In other words, using traditional math notation, we calculated $2x^2 + 5x + 3$ for $x = 6$.
This example shows that _decode_ can be used to calculate a polynomial represented by the coefficients of the unknown variable, sorted according to the decreasing powers of the variable.

For example, to calculate $3x^4 + 2x^2 - 7x + 2$ for $x = 1.2$, we can write

In [73]:
1.2 ⊥ 3 0 2 ¯7 2

2.7008

Do not forget zero coefficients for the missing powers of $x$ (here, we have no term in $x^3$).

This is equivalent to

In [74]:
(1.2 * 4 3 2 1 0) +.× 3 0 2 ¯7 2

2.7008

To calculate the value of a polynomial for several values of $x$, the values must be placed in a vertical 1-column matrix, to be compliant with the shape compatibility rules.
For example, to calculate the same polynomial for $x$ varying from 0 to 2 by steps of 0.2, we could write:

In [75]:
⎕← xs ← 0.2 × ¯1+⍳11

0 0.2 0.4 0.6 0.8 1 1.2 1.4 1.6 1.8 2

In [76]:
2⍕ (⍪xs) ⊥ 3 0 2 ¯7 2

2.00 0.68 ¯0.40 ¯1.09 ¯1.09 0.00 2.70 7.64 15.58 27.37 44.00

The results have been displayed with only 2 decimal digits using an appropriate _format_.

Finally, let us calculate $4x^4 + 2x^3 + 3x^2 - x - 6$ for $x = -1.5$:

In [77]:
¯1.5 ⊥ 4 2 3 ¯1 ¯6

15.75

Notice how we used a negative base.
That may seem strange, but there is nothing mathematically wrong with it.

(Mathematical-Functions-Right-aligning-Text)=
#### Right-aligning Text

We mentioned that _decode_ could be replaced by an _inner product_ using the following equivalence: `base ⊥ values ←→ weights +.× values`, if `weights ← ⌽1,×\⌽1↓base`.

What happens if the vector `base` contains one or more zeroes?

In [78]:
⌽1,×\⌽1↓10 8 5 3 10 2  ⍝ base has no zeroes

2400 300 60 20 2 1

In [79]:
⌽1,×\⌽1↓10 8 0 3 10 2  ⍝ base has a zero

0 0 60 20 2 1

We can see that the weights to the left of a zero are all forced to become zero.
In other words, `10 8 0 3 10 2 ⊥ values` is strictly identical to `0 0 0 3 10 2 ⊥ values`.

Let us use this discovery to right-align some text containing blanks:

In [80]:
⎕← text ← ↑'This little' 'text contains both' 'embedded and' 'trailing blanks'

This little 
text contains both
embedded and 
trailing blanks

In [81]:
text=' '

0 0 0 0 1 0 0 0 0 0 0 1 1 1 1 1 1 1
0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0
0 0 0 0 0 0 0 0 1 0 0 0 1 1 1 1 1 1
0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 1 1

If we use the rows of this Boolean matrix as decoding bases, the result above states that the ones placed to the left of a zero will not have any effect.
This means that the expression

In [82]:
¯1 + 1⊥⍨text=' '

7 0 6 3

counts the number of trailing `1`s in each row.
Thus, we can use those numbers to rotate right each row:

In [83]:
text⌽⍨1-1⊥⍨text=' '

This little
text contains both
 embedded and
 trailing blanks

Another example uses the same property of _decode_: given a Boolean vector `bin`, one can find how many `1`s there are to the right of the last zero using the expression `bin⊥bin`:

In [84]:
bin ← 0 1 1 0 1 1 1  ⍝ 3 trailing ones
bin⊥bin

3

(Mathematical-Functions-Randomised-Values)=
## Randomised Values

**Random numbers** are often used for demonstration purposes or to test an algorithm.
Strictly speaking, "random" would mean that a set of values is completely unpredictable.
This is not the case for numbers generated by a computer: they are, by definition, perfectly deterministic.

However, the values produced by an algorithm may appear to a human being as if they were random values when, given a subset of those numbers, the human is unable to predict the next values in the sequence.
If this first condition is satisfied, and if all of the unique values in a long series appear approximately the same number of times, these values can be qualified as _pseudo-random_ numbers.

In APL, the question mark `?` is used to produce pseudo-random numbers.

(Mathematical-Functions-Deal-Dyadic-Usage)=
### Deal: Dyadic Usage

The dyadic usage of the question mark is called _deal_.
The expression `number ? limit` produces as many unique pseudo-random integers as specified by `number`, all among `⍳limit`.
For this reason, we must have `number≤limit`.

The name of the function _deal_ relates to dealing cards.
When you are dealt a hand of cards from a deck, the cards that you get are arbitrary and you cannot be given the same card twice.
You may get cards with the same value or with the same suit, but you **cannot** get the exact same card twice.

Some examples follow.
First, we deal a hand of 7 numbers out of 52:

In [85]:
7?52

36 35 20 49 44 5 51

If we execute the same expression again, we get a different set of values:

In [86]:
7?52

50 26 10 40 4 34 2

If `number=limit`, we get a permutation of `⍳limit`:

In [87]:
12?12

6 9 12 4 10 7 1 2 5 8 11 3

And as we mentioned before, we cannot have `number>limit`:

In [88]:
13?12

DOMAIN ERROR: Deal right argument must be greater than or equal to the left argument
      13?12
        ∧


(Mathematical-Functions-Roll-Monadic-Use)=
### Roll: Monadic Use

The monadic use of the question mark is called _roll_.
The name relates to rolling a die.

(Mathematical-Functions-Positive-Integer-Arrays-as-Argument)=
#### Positive Integer Arrays as Argument

In the expression `?array`, where `array` is any array of positive integer values, each item `a` of `array` produces a pseudo-random value from `⍳a`, so that the result is an array of the same shape as `array`.
Each item of the result is calculated independently of the other items.

Here are some examples:

In [89]:
?1 2 3 4

1 2 3 4

In [90]:
?10 20 30 40

6 2 6 2

In [91]:
mat

VALUE ERROR: Undefined name: mat
      mat
      ∧


In [92]:
?mat

VALUE ERROR: Undefined name: mat
      ?mat
       ∧


If the argument of _roll_ is made of a repeated single value `v`, the result is an array made of values all taken from `⍳v`.
This makes it possible to produce any number of values within the same limits.

For example, we can simulate rolling a die 10 times:

In [93]:
?10⍴6

6 6 6 3 4 6 2 3 5 6

As another example, we can fill a matrix with the number `20`.
Then, for each of those numbers, we generate a pseudo-random number in `⍳20`:

In [94]:
?4 8⍴20

15 10 1 14 16 6 11 4
 1 19 5 8 12 19 2 17
 1 15 8 3 5 7 8 11
12 19 2 12 17 19 14 20

(Mathematical-Functions-0-as-Argument)=
#### 0 as Argument

The function _deal_ can also handle `0` as an argument (either as a scalar, or as element(s) of the argument array).
When presented with a zero, the function _deal_ will generate a pseudo-random decimal number strictly between 0 and 1:

In [95]:
?0

0.355175

In [96]:
?0

0.978111

In [97]:
?0

0.749925

In [98]:
?6 6 0 0 3 3

5 5 0.562462 0.548926 3 3

(Mathematical-Functions-Derived-Usages)=
### Derived Usages

(Mathematical-Functions-Random-Number-Within-an-Arbitrary-Range)=
#### Random Number Within an Arbitrary Range

We have seen that the values produced by _deal_ and _roll_ are always integer values extracted from `⍳v`, where `v` is a given limit.
However, it is possible to obtain a set of integers or decimal values between any limits, just by adding a constant and dividing by some appropriate value.

Imagine that we would like to obtain 10 decimal values, with 2 decimal digits, all between 743 and 761, inclusive.
We could follow these steps:

 - Let us first calculate integer values starting from 1:

In [99]:
lims ← 743 761
⎕← set ← ? 10⍴1 + 100×-⍨/lims

1494 514 1104 336 1586 444 1088 775 1459 1458

 - The values generated above are between 1 and 1801.
 - If we add `¯1+100×743` we obtain values between 74300 and 76100:

In [100]:
set +← ¯1+100×⌊/lims
⎕← set

75793 74813 75403 74635 75885 74743 75387 75074 75758 75757

 - Once divided by 100, they give decimal values with two decimal places between 743 and 761, inclusive:

In [101]:
set ÷← 100
set

757.93 748.13 754.03 746.35 758.85 747.43 753.87 750.74 757.58 757.57

We can check if the limits are well respected:

In [102]:
(⌊/,⌈/) set

746.35 758.85

(Mathematical-Functions-Sets-of-Random-Characters)=
#### Sets of Random Characters

Random characters can be obtained by indexing a set of characters by a random set of integer values smaller than or equal to the size of the character vector, as shown here:

In [103]:
⎕A[?3 15⍴≢⎕A]

CLOKIJUGZXBQPRW
BINSCHZCQSEYYHF
MMQLWFUMOBTKPJD

(Mathematical-Functions-Random-Link)=
### Random Link

The system variable `⎕RL` has been used throughout this book when creating large(r) arrays with random values and this section will explain what `⎕RL`, called _random link_, actually does.

(Mathematical-Functions-Reproducibility-When-Generating-Random-Numbers)=
#### Reproducibility When Generating Random Numbers

The primitive functions _deal_ and _roll_ do not produce really random numbers.
They produce pseudo-random numbers because there is actually an algorithm that runs to generate specific numbers, although the numbers generated by this algorithm "look random".

The algorithm makes use of something called a _seed value_, that you set within _random link_, and each time a new (pseudo-random) value is generated, this seed value is changed so that the next value will be different.

Sometimes, one might want to be able to recreate a random event later on.
For example, when running an experiment or a simulation, you may want to generate some numbers that you did not specify explicitly, but you may need to generate the exact same numbers later.
This is possible thanks to _random link_.

To reproduce a given set of random values, one can dynamically set `⎕RL` to any desired value, an integer in the range from 1 to 2147483646.

For example, throughout the book we have used the function _roll_ to generate a 4 by 6 matrix we have called `forecast`.
If we did not use _random link_, each time you opened the book, you would get a different matrix:

In [104]:
10×?4 6⍴55

350 40 80 190 460 190
250 380 50 410 180 150
550 70 170 310 380 370
520 390 470 300 540 250

In [105]:
10×?4 6⍴55

360 120 50 370 510 380
460 140 270 230 370 380
550 550 450 180 160 520
420 470 30 330 500 500

In [106]:
10×?4 6⍴55

320 410 220 20 270 550
270 220 350 500 360 310
130 70 260 180 430 100
450 250 40 540 350 310

To prevent this, before generating the matrix `forecast`, we set the _random link_ `⎕RL` to a specific value, which has been `⎕RL ← 73`.

In [107]:
⎕RL ← 73
⎕← forecast ← 10×?4 6⍴55

90 160 420 500 20 30
110 450 170 370 290 360
340 190 320 120 510 370
150 460 240 520 490 280

It is not relevant that the seed was set to `73`.
What is relevant is that the seed **was set to a fixed value**.

Each time you start the APL interpreter or clear the session with `)clear`, APL will reset the _random link_ so you can get pseudo-random values out of the box.
So, you only need to change the _random link_ when you know you will need reproducibility.

If you have set the seed for some computations but need to go back to generating pseudo-random numbers over which you have no control, you can set the _random link_ to _zilde_:

In [108]:
⎕RL ← ⍬

(Mathematical-Functions-The-Algorithm)=
#### The Algorithm

APL uses one of three algorithm to generate pseudo-random numbers and you can check which one is being used by looking at the value of `⎕RL`.
`⎕RL` is a 2-item vector, where the first item is the current seed and the second item is an integer between 0 and 2:

In [109]:
⎕RL

┌┬─┐
││1│
└┴─┘

In this case, the seed is `⍬` and the algorithm chosen is the algorithm 1.
If you read the Dyalog documentation, you will find some information on the three possible algorithms:

| Id | Name | Algorithm | Valid seed values |
| :- | :- | :- | :- |
| 0 | RNG0 | Lehmer linear congruential generator | `0`, `⍬`, or an integer in the range 1 to `¯2+2*31` |
| 1 | RNG1 | Mersenne Twister | `0`, `⍬`, an integer in the range 1 to `¯1+2*63`, or a 625-item integer vector |
| 2 | RNG2 | Operating System random number generator | `⍬` |

By default, Dyalog uses the Mersenne Twister (algorithm 1).

For the sake of curiosity, let us see how one would implement _roll_ using the algorithm 0:

In [110]:
∇ z ← Roll n
    seed ← (¯1+2*31)|seed×7*5
    z ← ⎕IO+⌊n×seed÷2*31
∇

The first instruction prepares the seed used for the next number.
_Residue_ is used to ensure that the result will always be smaller than `¯1+2*31`.
Consequently, `⎕RL÷2*31` always returns a value in the range 0-1.
Multiplied by the argument, it produces a result strictly smaller than `n`, and by adding `⎕IO` we get the desired result.

You can compare this implementation of `Roll` to the primitive function _roll_ if we set the seed and the algorithm to the correct values:

In [111]:
seed ← 73
Roll¨ 3⍴10

1 7 2

In [112]:
⎕RL ← 73 0
?3⍴10

1 7 2

In [113]:
seed

341854744

In [114]:
⊃⎕RL

341854744

(Mathematical-Functions-Some-More-Maths)=
## Some More Maths

(Mathematical-Functions-Logarithms)=
### Logarithms

The base `b` logarithm of a number `n` is calculated with the function _logarithm_ `⍟` that is typed with <kbd>APL</kbd>+<kbd>Shift</kbd>+<kbd>8</kbd>.

The value `l ← b⍟n` is such that `b*l` gives back the original number `n`.
Here are some examples:

In [115]:
10⍟1000

3

The logarithm of 1000 in base 10 is 3 because 10 to the 3rd power gives 1000:

In [116]:
10*3

1000

In [117]:
3⍟81

4

The base-3 logarithm of 81 is 4 because 3 to the power of 4 gives 81:

In [118]:
3*4

81

The monadic form of the function _logarithm_ gives the _natural logarithm_ (or _napierian logarithm_) of a number:

In [119]:
⍟10

2.30259

The base of the natural logarithm is also the base of the monadic function _power_, so we can easily compute it:

In [120]:
*1

2.71828

In [121]:
⍟*1

1

The monadic usage of the function _logarithm_ is the same as using `*1` as the left argument:

In [122]:
⍟10

2.30259

In [123]:
(*1)⍟10

2.30259

This is similar to how the monadic function _reciprocal_ `÷n` is equivalent to the dyadic usage `1÷n` or the monadic function _negate_ `-n` is equivalent to the dyadic usage `0-n`.

The relationship between the natural and the base `b` logarithms of a number is described the formulas below.
Given:

 - `n ← ⍟a`
 - `l ← b⍟a`
 - `e ← *1`, the base of the natural logarithm

Then, we have:

 - `n ≡ l×⍟b`
 - `l ≡ n×b⍟e`

(Mathematical-Functions-Factorial--Binomial)=
### Factorial & Binomial

The product of the first `n` positive integers, or the _factorial_ of `n`, is written as $n!$ in traditional mathematics.
APL uses the same symbol for the function, but in APL a monadic function is always placed to the left of its argument.
So, factorial looks like `!n` in APL.

As in mathematics, `!0` is equal to 1:

In [124]:
!0 1 2 3 4 5 6 7

1 1 2 6 24 120 720 5040

If `n` is a decimal number, `!n` gives the _gamma_ function of `n + 1`.
This is explained in {numref}`Mathematical-Functions-Gamma-and-Beta-Functions`.

The monadic function _factorial_ `!n` represents the number of possibilities when sorting `n` objects.
But if one picks only `p` objects among `n` objects, the number of possible _combinations_ is given by `(!n)÷(!p)×!n-p`.
This can be obtained directly using the dyadic function _binomial_ `p!n`.

For example, how many different hands can you be dealt if you are dealt 7 cards out of a 52 card deck?
That's

In [125]:
7!52

133784560

That is a surprisingly big number!
Over 130 million different hands.

If `p` is greater than `n`, then `p!n` gives the result 0.
In hindsight, this makes a lot of sense: if you have 10 items, in how many combinations can you make with 15 of those items?
Obviously, zero!
You do not even have enough items to consider doing those combinations.

The formula `(0,⍳n)!n` gives the coefficients of $(x + 1)^n$, which is the reason why the dyadic usage of `!` is called _binomial_.
One can obtain a set of coefficients with the following expression:

In [126]:
x ← ⍳5
⍉(0,x)∘.!x

1 1 0 0 0 0
1 2 1 0 0 0
1 3 3 1 0 0
1 4 6 4 1 0
1 5 10 10 5 1

The matrix above shows, for example, that $(x + 1)^3 = 1x^3 + 3x^2 + 3x + 1$.
The coefficients `1 3 3 1` were taken from the third row of the matrix.

(Mathematical-Functions-Trigonometry)=
### Trigonometry

(Mathematical-Functions-Multiples-of-π)=
#### Multiples of π

The mathematical constant pi (or π) is very useful in many mathematical and technical calculations.
It can be obtained via the primitive function _circle_ (the symbol `○`, typed with <kbd>APL</kbd>+<kbd>o</kbd>), which gives multiples of pi.
Do not confuse the function circle `o` with the jot `∘` used for the _outer product_ or the operators _beside_ and _bind_.

$\pi \times n$ can be obtained with `○n`:

In [127]:
○1 2 0.5

3.14159 6.28319 1.5708

In [128]:
○ ÷3

1.0472

This last expression gives $\frac\pi3$.
However, be careful!
The symbol `○`, by itself, does not represent the value π that we divided by 3 with `○÷3`.
`÷3` computed the reciprocal of 3 and then the monadic function _pi times_ got applied to that.

(Mathematical-Functions-Circular-and-Hyperbolic-Trigonometry)=
#### Circular and Hyperbolic Trigonometry

Using the dyadic form of the glyph _circle_, one can obtain all the possible direct and inverse functions of circular and hyperbolic trigonometry.

the trigonometric function is designated by the left argument to `○` according to the table below.
You can see that the positive left arguments refer to direct trigonometric functions, while negative arguments refer to their inverse functions.
Values from 1 to 3 calculate circular functions and values from 5 to 7 calculate hyperbolic functions.

(Mathematical-Functions-Direct-Trigonometric-Functions)=
##### Direct Trigonometric Functions

| `f` | `f ○ x` |
| :- | :- |
| `0` | $\sqrt{1 - x^2}$ |
| `1` | $\sin{x}$ |
| `2` | $\cos{x}$ |
| `3` | $\tan{x}$ |
| `4` | $\sqrt{1 + x^2}$ |
| `5` | $\text{sinh }x$ |
| `6` | $\text{cosh }x$ |
| `7` | $\text{tanh }x$ |


(Mathematical-Functions-Inverse-Trigonometric-Functions)=
#### Inverse Trigonometric Functions

| `f` | `f ○ x` |
| :- | :- |
| `¯1` | $\arcsin{x}$ |
| `¯2` | $\arccos{x}$ |
| `¯3` | $\arctan{x}$ |
| `¯4` | $\sqrt{-1 + x^2}$ |
| `¯5` | $\text{argsinh }x$ |
| `¯6` | $\text{argcosh }x$ |
| `¯7` | $\text{argtanh }x$ |

For example:

 - `2○x` means $\cos x$;
 - `5○x` means $\text{sinh }x$;
 - `¯2○x` means $\arccos x$;
 - `0○val` calculates $|\cos x|$ if $val = \sin x$, or $|\sin x|$ if $val = \cos x$;
 - `4○val` calculates $|\text{cosh }x|$ if $val = \text{sinh }x$; and
 - `¯4○val` calculates $|\text{sinh }x|$ if $val = \text{cosh }x$.

For direct circular trigonometry (`f∊1 2 3`) the value of `x` must be in radians.
For inverse circular trigonometry (`f∊-1 2 3`) the returned result is in radians.

(Mathematical-Functions-Concrete-Trigonometry-Examples)=
#### Concrete Trigonometry Examples

Let us compute the cosine of $0$, $\pi/6$, and $\pi/3$, respectively:

In [129]:
2○ 0,○÷6 3

1 0.866025 0.5

Next, we calculate three different functions for $\pi/3$:

In [130]:
1 2 3 ○ ○÷3

0.866025 0.5 1.73205

These three functions were, respectively, the sine, the cosine, and the tangent.

Now, we confirm that $\arctan 1 = \pi/4$:

In [131]:
(¯3○1) = ○÷4

1

We can also calculate the sine and cosine of $\pi/2$:

In [132]:
1 2 ○ ○÷2

1 6.12323E¯17

The very last result shows that the algorithms used to calculate the circular or hyperbolic values sometimes lead to very minor rounding approximation errors.
The second value is very close to zero, but it should be zero.

(Mathematical-Functions-Manipulating-Complex-Numbers)=
#### Manipulating Complex Numbers

Since the support for complex numbers was added to Dyalog APL, the function _circle_ was extended to provide some useful functions to manipulate complex numbers.
We list them in the table below, where `x` represents an arbitrary number.

| `f` | `f ○ x` |
| :- | :- |
| `8`| `(-1+x*2)*0.5` |
| `9`| real part of `x` |
| `10` | `\|x`, the magnitude of `x` |
| `11` | imaginary part of `x` |
| `12` | phase of `x` |
| `¯8` | `-8○x` |
| `¯9` | `x` |
| `¯10` | `+x`, the complex conjugate of `x` |
| `¯11` | `x×0J1` |
| `¯12` | `*x×0J1` |

To exemplify the usage of some of these left arguments, we will verify some simple identities.

For example, we can split a complex number up and put it back together:

In [133]:
c ← 3J4

In [134]:
re ← 9○c
im ← 11○c
c ≡ re+¯11○im

1

Using the function _magnitude_ directly should be the same as using `10○`, which should be the same as computing the magnitude of a complex number by hand:

In [135]:
⎕← mag ← |c

5

In [136]:
mag≡10○c

1

In [137]:
mag≡0.5*⍨+/re im*2

1

The phase of a complex number is given by the arctangent of the ratio of its imaginary and real parts:

In [138]:
⎕← phase ← 12○c

0.927295

In [139]:
phase≡¯3○im÷re

1

Finally, we can reconstruct a complex number from its phase and magnitude (which are used when representing complex numbers in polar form):

In [140]:
c ≡ magׯ12○phase

1

(Mathematical-Functions-GCD-and-LCM)=
### GCD and LCM

(Mathematical-Functions-Greatest-Common-Divisor-GCD)=
#### Greatest Common Divisor (GCD)

When applied to binary values, the symbol `∨` represents the Boolean function _or_.
The same symbol can be applied to numbers other than 0 and 1.
Then, it calculates their _greatest common divisor_, or GCD.
`∨` is always a dyadic scalar function: applied to arrays of the same shape, it gives a result of the same shape; applied between a scalar and any array, it gives a result of the same shape as the array, and the scalar value is reused as needed.

In [141]:
15180 ∨ 285285 ¯285285 47

165 165 1

The result of GCD is always positive.
Two integers always have, at least, the number 1 as a common divisor.
When the result of GCD between two integers is 1, that means there was no other divisor in common between the two integers and we say that those two integers are _coprime_.
For example, 15180 and 47 are coprime.

As always, if one of the items of the argument to `∨` is 0, the corresponding item from the other argument is returned (41, in the case below):

In [142]:
5180 0 28 ∨ 6142 41 19

74 41 1

The function _gcd_ also works with non-integer arguments:

In [143]:
5178 417 28 ∨ 7.4 0.9 1.4

0.2 0.3 1.4

(Mathematical-Functions-Lowest-Common-Multiple-LCM)=
#### Lowest Common Multiple (LCM)

When applied to binary values, the symbol `∧` represents the Boolean function _and_.
The same symbol can be applied to numbers other than 0 and 1.
Then, it calculates their lowest common multiple, or LCM.
`∧` is also a dyadic scalar function.

In [144]:
152 1 6183 ¯519 0 ∧ 316 8 411 24 16

12008 8 847071 ¯4152 0

The GCD and the LCM satisfy a relationship: if `a` and `b` are two numbers, then we have that `(a×b) ≡ (a∨b)×(a∧b)`.
This relationship can be rewritten as the elegant train `(×≡∨×∧)`, which always returns 1:

In [145]:
428 (×≡∨×∧) 561

1

In [146]:
¯28 (×≡∨×∧) 561

1

In [147]:
428 (×≡∨×∧) ¯51

1

In [148]:
¯42 (×≡∨×∧) ¯61

1

(Mathematical-Functions-Set-Union-and-Intersection)=
### Set Union and Intersection

Mathematical set theory describes the following two functions between two sets `a` and `b`:

 - _intersection_ `a ∩ b`, gives the items in both sets; and
 - _union_ `a ∪ b`, gives the items that are in either set.

The same functions are found in Dyalog, using the same symbols.
To type the symbol for set intersection `∩`, use <kbd>APL</kbd>+<kbd>c</kbd>.
To type the symbol for set union, use <kbd>APL</kbd>+<kbd>v</kbd>.
They work on scalar and vector arguments:

In [149]:
'Hey' 'give' 'me' 53 'dollars' ∩ 53 'euros' 'not' 'dollars'

┌──┬───────┐
│53│dollars│
└──┴───────┘

In [150]:
'Hey' 'give' 'me' 53 'dollars' ∪ 53 'euros' 'not' 'dollars'

┌───┬────┬──┬──┬───────┬─────┬───┐
│Hey│give│me│53│dollars│euros│not│
└───┴────┴──┴──┴───────┴─────┴───┘

In mathematics, we use these two functions with _sets_, which are not the same thing as APL vectors.
APL vectors are ordered and may contain duplicates, which makes a couple of conventions necessary.

(Mathematical-Functions-Intersection)=
#### Intersection

The result of the function _intersection_ is in the order that the items appear in the left argument, including duplicates.

In [151]:
1 1 2 3 2 3 ∩ 2 3

2 3 2 3

In [152]:
text ← 'The quick brown fox jumps over the lazy dog'
vowels ← 'aeiouy'
text ∩ vowels

euioouoeeayo

In fact, the result of the _intersection_ is equal to the left argument, but with all items that are not found in the right argument removed.
aWe can check this by using the function _without_ twice:

In [153]:
(text∩vowels) ≡ text~text~vowels

1

(Mathematical-Functions-Union)=
#### Union

The result of the function _union_ is always the left argument, followed by all items of the right argument that are not already found in the left argument – including duplicates:

In [154]:
1 1 2 3 2 3 ∪ 2 2 2 4 4 4 6 6 6

1 1 2 3 2 3 4 4 4 6 6 6

Again, we can verify this property by making use of the function _without_:

In [155]:
left ← 1 1 2 3 2 3
right ← 2 2 2 4 4 4 6 6 6
(left∪right) ≡ left,right~left

1

(Mathematical-Functions-Unique)=
#### Unique

The function _union_ can also be used monadically, in which case it is the function _unique_.
The function _unique_ works on arrays of arbitrary dimension and returns an array of the same rank, with all the unique major cells of the initial argument, in order of appearance:

In [156]:
∪1 1 2 3 2 3

1 2 3

In [157]:
⎕← months ← ↑'Jan' 'Feb' 'Jan' 'Apr' 'Dec' 'Apr' 'Mar'

Jan
Feb
Jan
Apr
Dec
Apr
Mar

In [158]:
∪months

Jan
Feb
Apr
Dec
Mar

(Mathematical-Functions-Domino)=
## Domino

(Mathematical-Functions-Some-Definitions)=
### Some Definitions

(Mathematical-Functions-Identity-Matrix)=
#### Identity Matrix

Any number multiplied by 1 stays unchanged.
Similarly, a matrix multiplied by a special square Boolean matrix remains unchanged.
We call _identity matrix_ to that special matrix.

The _identity matrix_ is a square matrix that is all zeroes except in the main diagonal, where it has ones.

For example, if we have a 3 by 3 matrix

In [159]:
⎕← m ← 3 3⍴12 50 7 44 3 25 30 71 80

12 50 7
44 3 25
30 71 80

Then, the identity matrix for `m` must be a 3 by 3 matrix of zeroes with ones in the main diagonal.
A simple way of creating the identity matrix of size `n` is by using the expression `n n⍴n+1↑1`:

In [160]:
⎕← i ← 3 3⍴4↑1

1 0 0
0 1 0
0 0 1

Now, if we multiply `m` and `i` together, we get `m` back:

In [161]:
m+.×i

12 50 7
44 3 25
30 71 80

In [162]:
i+.×m

12 50 7
44 3 25
30 71 80

(Mathematical-Functions-Inverse-Matrices)=
#### Inverse Matrices

If we multiply 4 by 0.25, or 0.25 by 4, we obtain 1, which is the identity item for multiplication.
Alternatively, we can say that 0.25 is the reciprocal, or inverse, of 4, and vice-versa.

Given that `i` is the identity item for matrix multiplication, if we can find two matrices `m` and `m_` whose product is `i`, we can say that those two matrices are the inverse of each other.

For example:

In [163]:
⎕← m ← 3 3⍴1 0 2 0 2 1 .5 3 1.5

1 0 2
0 2 1
0.5 3 1.5

In [164]:
⎕← m_ ← 3 3⍴0 ¯3 2 ¯.25 ¯.25 .5 .5 1.5 ¯1

0 ¯3 2
¯0.25 ¯0.25 0.5
 0.5 1.5 ¯1

If we multiply these two matrices together, in either order, we get the identity matrix:

In [165]:
m+.×m_

1 0 0
0 1 0
0 0 1

In [166]:
m_+.×m

1 0 0
0 1 0
0 0 1

Note that, in general, `a+.×b` and `b+.×a` give different results!

Here is a second example with 2 by 2 matrices:

In [167]:
⎕← m ← 2 2⍴2 1 4 1

2 1
4 1

In [168]:
⎕← m_ ← 2 2⍴¯.5 .5 2 ¯1

¯0.5 0.5
 2 ¯1

In [169]:
m+.×m_

1 0
0 1

Now, the result is the identity matrix of size 2, because the matrices `m` and `m_` are 2 by 2 (and not 3 by 3).

For the moment, we will only concern ourselves with inverses for square matrices.
We shall in {numref}`Mathematical-Functions-Domino-and-Rectangular-Matrices` that it is also possible to define inverses for non-square matrices.

(Mathematical-Functions-Matrix-Inverse)=
### Matrix Inverse

(Mathematical-Functions-Monadic-Domino)=
#### Monadic Domino

APL provides a _matrix inverse_ primitive function, represented by the symbol `⌹` and typed with <kbd>APL</kbd>+<kbd>Shift</kbd>+<kbd>=</kbd>, which is the same key as `÷`.
Because of its appearance, this symbol is named _domino_.

The monadic usage of _domino_ returns the inverse of a matrix:

In [170]:
⌹m

¯0.5 0.5
 2 ¯1

Calculating the inverse of a matrix is a complex operation, and the result might be only an approximation.
For example, the inverse of the matrix `m` has the value `¯1` in the bottom right corner.
However, it would not be surprising if the function _matrix inverse_ had returned this result, instead:

In [171]:
2 2⍴¯.5 .5 2 ¯.999999

¯0.5 0.5
 2 ¯0.999999

(Mathematical-Functions-Singular-Matrices)=
#### Singular Matrices

In normal arithmetic, zero has no inverse and `÷0` in APL causes a `DOMAIN ERROR`.

In the same way, some matrices cannot be inverted.
Those matrices are said to be _singular_:

In [172]:
⌹ ⎕← 3 3⍴1 3 5 3 4 15 2 7 10

1 3  5
3 4 15
2 7 10
DOMAIN ERROR
      ⌹⎕←3 3⍴1 3 5 3 4 15 2 7 10
      ∧


(Mathematical-Functions-Solving-a-Set-of-Linear-Equations)=
#### Solving a Set of Linear Equations

Here is a set of three linear equations with three unknowns, written in traditional mathematical notation:

$$
\begin{alignat}{4}
-8 &= 3&&x + 2&&y -\ &&z \\
19 &= &&x -\ &&y + 3&&z \\
0 &= 5&&x + 2&&y
\end{alignat}
$$

This set of equations can be represented using a vector for the constants and a matrix for the coefficients of the three unknowns, as shown below:

In [173]:
cons ← ¯8 19 0

In [174]:
⎕← coefs ← 3 3⍴3 2 ¯1 1 ¯1 3 5 2 0

3 2 ¯1
1 ¯1 3
5 2 0

To solve the set of equations above, we must find a 3-element vector `xyz` such that `cons ≡ coefs +.× xyz`.

We can find such a vector provided that the matrix `coefs` has an inverse, i.e., that it is not singular.
Let us multiply both sides of the equality `cons ≡ coefs +.× xyz` by the inverse of `coefs`:

```APL
                cons ≡              coefs +.× xyz ←→
←→ (⌹coefs) +.× cons ≡ (⌹coefs) +.× coefs +.× xyz
```

Now, because we know that `id ← (⌹coefs) +.× coefs` is the identity matrix, we can simplify the expression:

```APL
   (⌹coefs) +.× cons ≡ (⌹coefs) +.× coefs +.× xyz ←→
←→ (⌹coefs) +.× cons ≡ id +.× xyz ←→
←→ (⌹coefs) +.× cons ≡ xyz
```

Eureka!
We found a way of calculating the values that we had to find:

In [175]:
⎕← xyz ← (⌹coefs) +.× cons

2 ¯5 4

In general, the solution to a set of linear equations is given by `solution ← (⌹coefficients) +.× constants`.

Note that, in the formula above, we multiply `constants` by the inverse (or reciprocal) of a matrix.
When dealing with numbers, multiplying by the reciprocal of a number is usually known as division, which motivates the name of the dyadic usage of the _domino_ `⌹`.

(Mathematical-Functions-Matrix-Division)=
### Matrix Division

The dyadic form of _domino_ is called _matrix divide_, so it can do exactly what we have just done: it can easily sets of linear equations like the one shown above.
We solve that set of linear equations again, this time using _matrix divide_:

In [176]:
cons⌹coefs

2 ¯5 4

Naturally, this method works only if the coefficient matrix has an inverse.
In other words, the set of equations must have a single solution.
If three is no solution, a `DOMAIN ERROR` will be reported.

We can summarise this as follows: given a system of `n` linear equations with `n` unknowns, where the matrix of coefficients is called `coefs` and the vector of constants is called `cons`, the solution `sol` of the system is given by `sol ← cons⌹coefs`.

(Mathematical-Functions-Two-or-Three-Steps-in-Geometry)=
### Two or Three Steps in Geometry

(Mathematical-Functions-A-Complex-Solution-to-a-Simple-Problem)=
#### A Complex Solution to a Simple Problem

To begin with, we invite you to study a complicated method to solve a simple problem.
Our intention is then to generalise this method to develop a solution for an everyday problem in statistical studies.

The goal is to find the coefficients of a straight line passing through two points `P` and `Q`, of which the coordinates are given below:

In [177]:
xs ← 2 4  ⍝ x coordinates of P and Q
ys ← 2 3  ⍝ y coordinates of P and Q

The points `P` and `Q` are shown in {numref}`fig-Graph_P_Q`.

(fig-Graph_P_Q)=
```{figure} ../res/Graph_P_Q.png
---
name: Graph_P_Q
---
A graph with the points `P` and `Q`.
```

The general equation describing a straight line is $y = mx + b$.
With our two points given, the following is obtained:

$$
\begin{align}
2 &= 2m + b
3 &= 4m + b
\end{align}
$$

This is a set of two linear equations in which the unknowns are $m$ and $b$.
Let us solve this set of equations by the method demonstrated in the previous section:

In [178]:
cons ← ys

In [179]:
⎕← coefs ← xs,[1.5]1

2 1
4 1

Now, $m$ and $b$ can be calculated using the method we saw above:

In [180]:
⎕← c ← cons⌹coefs

0.5 1

We can also compute the coefficients of the line directly from the vectors `xs` and `ys`:

In [181]:
⎕← c ← ys⌹xs,[1.5]1

0.5 1

In traditional notation, this gives us the equation $y = 0.5x + 1$ for the line.
Did you find this tedious?
Maybe, but now let us discover its scope.

(Mathematical-Functions-Calculating-Additional-Y-Coordinates)=
#### Calculating Additional Y Coordinates

Having found the coefficients of the line shown in the previous section, let us try to calculate the y coordinates of several points for which the x coordinates are known.

The coefficients of our line were obtained by this calculation:

In [182]:
c ← ys⌹coefs

We saw earlier that this is strictly equivalent to `c ← (⌹coefs) +.× ys`.
Let us multiply both terms of this expression by `coefs`:

```APL
             c ≡ (⌹coefs) +.× ys ←→
←→ coefs +.× c ≡ coefs +.× (⌹coefs) +.× ys ←→
←→ coefs +.× c ≡ ys
```

This shows that the y coordinates `ys` of some points placed on a line defined by the coefficients `c` can be calculated from their x coordinates `xs` by the formula `coefs +.× c` or, in a more explicit form: `ys ← c +.×⍨ xs,[1.5]1`.

Let us apply this technique to a set of points:

In [183]:
c +.×⍨ 0 ¯2 3 6,[1.5]1

1 0 2.5 4

You can check these values in {numref}`fig-Graph_P_Q`.

(Mathematical-Functions-Least-Squares-Fitting)=
### Least Squares Fitting

(Mathematical-Functions-Linear-Regression)=
#### Linear Regression

Our line was defined by two points.
What happens if we no longer have 2 points, but many?
Of course, there is a high probability that these points are not aligned.

As an example, suppose that we have twelve employees and we know their ages (in years) and salaries (in peanuts):

In [184]:
age ← 20 21 28 31 33 34 36 37 40 44 45 51
salaries ← 3071 2997 2442 3589 3774 3071 3108 5291 5180 7548 5772 5883

In [185]:
age,[.5]salaries

20 21 28 31 33 34 36 37 40 44 45 51
3071 2997 2442 3589 3774 3071 3108 5291 5180 7548 5772 5883

We shall place the ages on the x axis and salaries on the y axis (see {numref}`fig-Graph_Age_Salary`).

This time, the matrix

In [186]:
age,[1.5]1

20 1
21 1
28 1
31 1
33 1
34 1
36 1
37 1
40 1
44 1
45 1
51 1

will have more rows (12) than columns (2), and the set of equations has no solution, which confirms that no single straight line can join all those points.

In this type of situation, it may be desirable to define a straight line which best represents the spread of points.
Generally, a straight line is sought such that the sum of the squares of the deviations of the y coordinates between the given points and the line is minimised.
This particular line is called the _least squares line_ and the process of finding it is called _linear regression_.

To find this line, we shall use `⌹` once more.
The expression used to calculate the coefficients of a line passing through two points can be applied to a rectangular matrix, and the function _matrix divide_ gives the coefficients of the _least squares line_ passing through the set of points.
Is it not magic?

For the given points, here is the calculation:

In [187]:
⎕← c ← salaries⌹age,[1.5]1

134.946 ¯412.6

This gives the line $y = 134.946x - 412.6$.

We can calculate the approximate salaries located on the line, at the same x coordinates as the given points and compare with the actual salaries:

In [188]:
0⍕ salaries ,[.5] c+.×⍨age,[1.5]1

3071 2997 2442 3589 3774 3071 3108 5291 5180 7548 5772 5883
 2286 2421 3366 3771 4041 4176 4445 4580 4985 5525 5660 6470

(fig-Graph_Age_Salary)=
```{figure} ../res/Graph_Age_Salary.png
---
name: Graph_Age_Salary
---
Graph with the salaries (in peanuts) as a function of the age
```


(Mathematical-Functions-Extension)=
#### Extension

In the previous example, we measured the effect of a single factor (age) on a single observation (salary) using a linear model.

What if we want to use the same model to explore the relationship between several factors and a single observation?
The following example is inspired by a controller in IBM France who tried to see if the heads of his commercial agencies had "reasonable" expense claim forms.

The amounts were stored in a vector `amounts`.
He tried to measure the effect of 4 factors on these expenses amounts:

 - the number of salesmen in each agency, `nbmen`;
 - the size of the area covered by each agency, `radius`;
 - the number of customers in each agency, `nbcus`; and
 - the annual income produced by each agency, `income`.

In other words, he tried to find the vector `tc` of 5 theoretical coefficients, $c_1, \cdots, c_5$, which most closely satisfy this equation:

$$
{\rm amounts} = {\rm nbmen}\times c_1 + {\rm radius}\times c_2 + {\rm nbcus}\times c_3 + {\rm income}\times c_4 + c_5
$$

Or, in APL, we want to find the 5-element vector `tc` that most closely satisfies this equality:

```APL
amounts ≡ tc +.× nbmen radius nbcus income 1
```

Here is the data:

In [189]:
amounts ← 40420 23000 28110 32460 25800 33610 61520 44970
nbmen ← 25 20 24 28 14 8 31 17
radius ← 90 50 12 12 30 30 120 75
nbcus ← 430 87 72 210 144 91 207 161
income ← 2400 9000 9500 4100 6500 3300 9800 4900

Let us apply exactly what we did on ages and salaries, and calculate the following variables:

In [190]:
⎕← coefs ← nbmen,radius,nbcus,income,[1.5]1

25 90 430 2400 1
20 50 87 9000 1
24 12 72 9500 1
28 12 210 4100 1
14 30 144 6500 1
 8 30 91 3300 1
31 120 207 9800 1
17 75 161 4900 1

Next, we compute the 5 coefficients of the least squares line:

In [191]:
⎕← 2⍕tc ← amounts⌹coefs

1154.23 362.14 ¯99.39 ¯3.33 31193.65

Now, we compute the y coordinates of the points on the least squares line:

In [192]:
ys ← coefs+.×tc

Finally, we compute the differences between the expenses predicted by the expenses claims and the actual claims, and we compute the percentage deviation:

In [193]:
diff ← amounts-ys
pcent ← 100×diff÷ys

Finally, we display all the data, with a "-" next to the agencies for which the expenses are significantly above the predicted value (possibly because the agencies have bad managers) and a "+" sign next to the agencies for which the expenses are significantly below the predicted value (possibly because the agencies have good managers):

In [194]:
⎕← flag ← '+ -'[1+¯10 10⍸pcent]

+- -

In [195]:
⎕← title ← 29↑'   Real Normal   Diff      %'

Real Normal Diff %

In [196]:
title⍪(7 0⍕amounts,ys,diff,[1.5]pcent),flag

Real Normal Diff % 
 40420 41914 ¯1494 ¯4 
 23000 33773 ¯10773 ¯32+
 28110 24455 3655 15-
 32460 33335 ¯875 ¯3 
 25800 22264 3536 16-
 33610 31260 2350 8 
 61520 57229 4291 7 
 44970 45660 ¯690 ¯2

(Mathematical-Functions-Non-linear-Adjustment)=
#### Non-linear Adjustment

In this last example we used independent factors and tried to combine them with a linear expression.
We could as well have used vectors linked one to the other by any mathematical expression, like `results ← (c1×var)+(c2×var*2)+(c3×⍟var)+c4` (if this makes sense).
A typical case is trying to fit a set of points with a polynomial curve.
Here are 8 points:

In [197]:
xs ← ¯1 ¯1 0.5  1.5  2 2 3 4
ys ← ¯3 ¯1 0   ¯1   ¯1 1 3 5

A linear regression would give a line with the following coefficients:

In [198]:
2⍕ ys⌹xs,[1.5]1

1.22 ¯1.31

The right argument was obtained by laminating 1 to `xs`.
However, we could just as well have obtained it with the following _outer product_:

In [199]:
xs∘.*1 0

¯1 1
¯1 1
 0.5 1
 1.5 1
 2 1
 2 1
 3 1
 4 1

Now, instead of taking only powers 1 and 0 of `xs`, we could extend the scope of the powers up to the third degree, for example:

In [200]:
xs∘.*3 2 1 0

¯1 1 ¯1 1
¯1 1 ¯1 1
 0.125 0.25 0.5 1
 3.375 2.25 1.5 1
 8 4 2 1
 8 4 2 1
27 9 3 1
64 16 4 1

We would then obtain not the coefficients of a straight line, but those of a third degree polynomial curve, shown in {numref}`fig-Graph_Cubic`.

(fig-Graph_Cubic)=
```{figure} ../res/Graph_Cubic.png
---
name: Graph_Cubic
---
A third degree polynomial adjusted to some points.
```

To get the coefficients, we do the usual computation:

In [201]:
2⍕ c←ys⌹xs∘.*3 2 1 0

0.10 ¯0.16 0.58 ¯1.07

In other words, this set of points can be approximated by the line $0.1x^3 - 0.16x^2 + 0.58x - 1.07$.

(Mathematical-Functions-Exercises)=
## Exercises

```{exercise}
:label: ex-encode
Can you predict (and explain) the results of the two expressions below?
```


In [202]:
0 ⊥ 12 34 60 77 19

19

In [203]:
1 ⊥ 12 34 60 77 19

202

```{exercise}
:label: ex-hex
In a binary number system, a group of 4 bits represents values from 0 to 15.
Those 4 bits represent the 16 states of a base 16 number system known as the hexadecimal number system.
This system is very often used in computer science, with the numbers 0-9 represented by the characters "0"-"9", and the numbers 10-15 represented by the characters "A" to "F".
Write a function to convert hexadecimal values to decimal, and the reverse.
Assume hexadecimal values are represented as 4 character vectors:
```


In [204]:
D2H ← {'0123456789ABCDEF'[1+16 16 16 16⊤⍵]}

In [205]:
(⍪,D2H¨) ¯1+⍳16

┌──┬────┐
│0 │0000│
├──┼────┤
│1 │0001│
├──┼────┤
│2 │0002│
├──┼────┤
│3 │0003│
├──┼────┤
│4 │0004│
├──┼────┤
│5 │0005│
├──┼────┤
│6 │0006│
├──┼────┤
│7 │0007│
├──┼────┤
│8 │0008│
├──┼────┤
│9 │0009│
├──┼────┤
│10│000A│
├──┼────┤
│11│000B│
├──┼────┤
│12│000C│
├──┼────┤
│13│000D│
├──┼────┤
│14│000E│
├──┼────┤
│15│000F│
└──┴────┘

In [206]:
D2H¨ 6748 49679 60281

┌────┬────┬────┐
│1A5C│C20F│EB79│
└────┴────┴────┘

In [207]:
H2D ← {16⊥¯1+'0123456789ABCDEF'⍳⍵}

In [208]:
H2D¨ '1A5C' 'C20F' 'EB79'

6748 49679 60281

```{exercise}
:label: ex-sparse-array
A sparse array is an array of zeroes with only a small number of non-zero elements.
To save memory, sparse arrays are often represented in condensed forms, where we only store the positions of the non-zero values.
In this exercise, you will write two functions, `Contract` and `Expand`, to convert a sparse array into its compressed form and vice-versa.

`Contract` accepts an arbitrary sparse array `sa` as its only argument and returns a vector with the following characteristics:
 - the first element is the rank `r` of `sa`;
 - the next `r` elements are the shape of `sa`;
 - the next `n` elements are the `n` non-zero values of `sa`; and
 - the final `n` elements are the positions of the corresponding non-zero elements. However, the final `n` elements should be the indices of the non-zero values in the ravel of `sa`, and not the high-rank indices.

The function `Expand` should do the reverse operation.

Implement `Contract` and `Expand` without using the function _ravel_.
```

Here is a worked example with a 5 by 7 matrix:

In [209]:
sa ← 5 7⍴0 ⋄ sa[1;2] ← 3 ⋄ sa[4;1] ← 5.6
⎕← sa

0 3 0 0 0 0 0
0 0 0 0 0 0 0
0 0 0 0 0 0 0
5.6 0 0 0 0 0 0
0 0 0 0 0 0 0

In [210]:
]dinput
Contract ← {
    flat ← 1+(⍴⍵)⊥¯1+⍉↑idx←⍸0≠sa
    (≢⍴⍵),(⍴⍵),⍵[idx],flat
}

In [211]:
Contract sa

2 5 7 3 5.6 2 22

The matrix `sa` has two non-zero values at indices `1 2` and `4 1`, respectively.
However, in the ravel of `sa`, those two non-zero values are at indices `2` and `22`:

In [212]:
(,sa)[2 22]

3 5.6

Thus, the final part of the result of `Contract sa` should be the 4-item vector `3 5.6 2 22`:

```APL
      Contract sa
2 5 7 3 5.6 2 22
↑ ~~~ ∧∧∧∧∧ ∧∧∧∧
| |   |   the corresponding indices are 2 and 22
| |   the non-zero values are 3 and 5.6
| shape of sa
rank of sa
```

The function `Expand` should do the reverse operation:

```{exercise}
:label: ex-random
Create three variables, filled with random integers, according to the following specifications:

 - a vector of 12 values between 8 and 30, without duplicates;
 - a 4 by 6 matrix filled with values between 37 and 47, with possible duplicates; and
 - a 5 by 2 matrix filled with values between ¯5 and 5, without duplicates.
```


```{exercise}
:label: ex-random-decimals
Create a vector of 15 random numbers between 0.01 and 0.09 inclusive, with 3 significant digits each, with possible duplicates.
```


```{exercise}
:label: ex-read-random
What will we obtain by executing the expression `10+?(10+?10)⍴10`?
```


```{exercise}
:label: ex-random-choose
We would like to obtain a vector of 5 items, chosen randomly and without duplicates among the elements of `list ← 12 29 5 44 31 60 8 86`.
```


```{exercise}
:label: ex-random-length
Create a vector with a length randomly chosen between 6 and 16, and filled with random integers between 3 and 40 inclusive, with possible duplicates.
```


```{exercise}
:label: ex-cos-approximation
The value of $\cos x$ can be calculated by the following formula, written with traditional mathematical notation:

$$
\cos x = x^0/0! - x^2/2! + x^4/4! - x^6/6! + x^8/8! - x^{10}/10! \cdots
$$

Write a function `n Cos x` that accepts an integer `n` as left argument and a value `x` as right argument and computes the value of the expression above up to the power `2×n`, for the value `x`.

(To verify your answer, `n Cos x` should be close to `2○x`.)
```


```{exercise}
:label: ex-circle
Try to evaluate the following expressions, and then check your result on the computer:

 - `(1○○÷4)*2`; and
 - `2×0.5+¯2○1○0.5`.

(This exercise assumes you have some trigonometry knowledge.)
```


```{exercise}
:label: ex-solve-linear-eqs
Find the solution to this set of equations:

$$
\begin{align}
x - y &= 5 \\
y - 2z &= -7 \\
z - x &= 2
\end{align}
$$
```


```{exercise}
:label: ex-linear-eqs-compute
Three variables $a$, $b$, and $c$, meet the following conditions:

$$
\begin{align}
a - b + 3c &= 13 \\
4b - 2a &= -6 \\
a - 2b + 2c &= 10
\end{align}
$$

Can you calculate the value of $3a + 5b - c$?
```


(Mathematical-Functions-The-Specialists-Section)=
## The Specialist's Section

<br />
<center><i>Each chapter is followed by a "Specialist's Section" like this one. This section is dedicated to skilled APLers who wish to improve their knowledge.</i>
    
You will find here rare or complex usages of the concepts presented in this chapter, or discover extended explanations which need the knowledge of some symbols that will be seen much further in the book.
    
<b>If you are exploring APL for the first time, skip this section and go to the next chapter.</b></center>

(Mathematical-Functions-Advanced-Encode-and-Decode)=
### Advanced Encode and Decode

(Mathematical-Functions-Special-Decoding)=
#### Special Decoding

If you replace _decode_ by the equivalent _inner product_, you can easily check these surprising properties:

```APL
0⊥v ←→ ⍬⍴¯1↑v
1⊥v ←→ +/v
```

When dealing with integers, the values are normally smaller than the corresponding items of the _base_.
For example, it would be unusual to say that a time is equal to 7 hours, 83 minutes, and 127 seconds.
However, _decode_ works perfectly on such values:

In [213]:
24 60 60 ⊥ 7 83 127

30307

And, though we did not mention it, _encode_ (like _decode_) accepts decimal and/or negative bases:

In [214]:
3⍴5.2 ⊤ 160.23

4.23 4.23 4.23

(Mathematical-Functions-Processing-Negative-Values)=
#### Processing Negative Values

Imagine that we want to encode some numbers in any base, with 6 digits.
We will choose base 10, so that the result is readily understandable:

In [215]:
(6⍴10)⊤17

0 0 0 0 1 7

Now, let us try to predict the result of `(6⍴10)⊤¯17`.
Knowing that `17+¯17` is 0, it would be reasonable to expect that `((6⍴10)⊤17)+((6⍴10)⊤¯17)` also returns a result that is "zero" in some way – for example `6⍴0`?

In other words, we are looking for a 6-item vector that, when added to `0 0 0 0 1 7`, gives `6⍴0`.
The real challenge is that the items of the vector must only contain the digits 0 through 9!

Let us see what happens:

In [216]:
(6⍴10)⊤¯17

9 9 9 9 8 3

These results may seem a bit surprising.
In base 10, when the sum of two digits exceeds 9, a carry is produced, which is used when adding the next digits to the left, and so on.
When adding the encoded values of `17` and `¯17`, here is how the values are processed:

```
  1 1 1 1 1 1
    0 0 0 0 1 7
+   9 9 9 9 8 3
  -------------
= 1 0 0 0 0 0 0
```

If only 6 digits are kept, as agreed, you can see that the result is composed only of zeroes (the six leftmost digits in the result).

That would be the same for `431` and `¯431`:

In [217]:
(6⍴10)⊤431

0 0 0 4 3 1

In [218]:
(6⍴10)⊤¯431

9 9 9 5 6 9

We can observe the same behaviour in any other base.
For example, in base 5:

In [219]:
(6⍴5)⊤68

0 0 0 2 3 3

In [220]:
(6⍴5)⊤¯68

4 4 4 2 1 2

Remember that, in base 5, `0=3+2`!

The rule remains true for decimal values:

In [221]:
(6⍴10)⊤15.8

0 0 0 0 1 5.8

In [222]:
(6⍴10)⊤¯15.8

9 9 9 9 8 4.2

This may lead to a few frustrations during decoding if precautions are not taken:

In [223]:
10⊥(6⍴10)⊤ 143  ⍝ as expected

143

In [224]:
10⊥(6⍴10)⊤¯143  ⍝ this may seem wrong

999857

The result obtained from the two decodings is such that their sum is 1,000,000.

If one needs to encode and later decode values among which some may be negative, it is advisable to provide one more digit than is necessary, and to test its value.
In many computers, the internal representation of positive numbers begin with a 0-bit and negative numbers start with a 1-bit.
The same principle applies to the decoded values.
In base 10, positive values begin with the digit 0, and negative ones with a 9.
From this, we can deduce a general function to decode values encoded in a uniform base:

In [225]:
]dinput
Decode ← {  ⍝ base Decode values
    z ← ⍺⊥⍵
    p ← 1↑⍴⍵
    z-(⍺*p)×(z>¯1+⍺*p-1)
}

In [226]:
test ← (6⍴10) ⊤ ¯417.42 26 32 ¯1654 0 3.7 ¯55

In [227]:
10⊥test  ⍝ negative values are wrongly decoded...

999583 26 32 998346 0 3.7 999945

In [228]:
10 Decode test  ⍝ positive values are correctly decoded!

¯417.42 26 32 ¯1654 0 3.7 ¯55

(Mathematical-Functions-Multiple-EncodingDecoding)=
#### Multiple Encoding/Decoding

It is possible to decode a value using several bases simultaneously:

In [229]:
⎕← bb ← ⍉4 3⍴5 10 6

5 5 5 5
10 10 10 10
 6 6 6 6

The different bases are placed in different rows of the matrix.
Now, we can interpret `4 1 3 2` in 3 different bases:

In [230]:
bb⊥4 1 3 2

542 4132 920

In this very specific case, the 3 bases were all uniform; a simple column vector with one digit per row would have given the same conversion:

In [231]:
(⍪5 10 6)⊥4 1 3 2

542 4132 920

That is because a scalar base expands into as many digits as needed.

Similarly, it is possible to see how a single number could be represented in several bases:

In [232]:
(5 3⍴5 10 2)⊤27

0 0 1
0 0 1
1 0 0
0 2 1
2 7 1

```{admonition} Remark 
:class: tip
The arguments of _encode_ and _decode_ must obey the following rules:

- In `r ← bases ⊥ mat`, `⍴r` is equal to `(¯1↓⍴,bases),1↓⍴mat`.
- In `r ← bases ⊤ mat`, `⍴r` is equal to `bases ,⍥⍴ mat`.
```

Let us convert (_encode_) the following matrix `seconds` into hours, minutes, and seconds:

In [233]:
⎕← seconds ← 2 3⍴1341 5000 345 3600 781 90

1341 5000 345
3600 781 90

In [234]:
24 60 60 ⊤ seconds

0 1 0
 1 0 0

22 23 5
 0 13 1

21 20 45
 0 1 30

The first plane of the 3D array contains the hours, the second plane contains the minutes, and the third plane contains the seconds.

(Mathematical-Functions-Gamma-and-Beta-Functions)=
### Gamma and Beta Functions

In mathematics, the _gamma_ function of a variable $x$ is defined by the following formula:

$$
\Gamma(x) = \int_0^\infty t^{x-1}e^{-t} dt
$$

This function satisfies the recurrence relationship $\Gamma(x) = x\times \Gamma(x-1)$.
If $x$ is an integer, $\Gamma(x)$ is the same as $(x-1)!$.
In APL, monadic `!n` gives the _gamma_ function of `n+1`:

In [235]:
! 2.4 3 3.2 3.4

2.98121 6 7.75669 10.1361

The dyadic form of `!` gives the _binomial_ function, which satisfies the following identity with the mathematical _beta_ function: $\beta(a, b)$ is identical to `÷b×(a-1)!a+b-1`.

(Mathematical-Functions-Domino-and-Rectangular-Matrices)=
### Domino and Rectangular Matrices

We have defined the _matrix inverse_ for square matrices, but throughout this chapter we have used the function _matrix divide_ on rectangular matrices.
In the following sections we shall define what are the "left inverse" and the "right inverse" of a given rectangular matrix.

```{admonition} Remark 
:class: tip
Under the hood, _domino_ uses Householder transformations, a technique based on the Lawson & Hanson algorithm, an extension of the Golub & Businger algorithm.
```


(Mathematical-Functions-Left-Inverse-of-a-Matrix)=
#### Left Inverse of a Matrix

We have been working in this context:

 - `ys` was a vector of values we wanted to predict.
 - `coefs` was a rectangular matrix containing a set of columns containing variables (for example, number of salesmen, size of the area, number of customers, ...).
 - `c` contained the coefficients of the least squares line. They should be such that the sum of the squares of the differences between points on this line and the given points in `ys` is minimal.

In mathematics, three major properties can be demonstrated:

 - The coefficients `c` can be obtained from `ys` by means of a linear regression or, in other words, an _inner product_ involving a matrix `inv`: `c ← inv+.×ys`.
 - This matrix `inv` is unique.
 - It can be obtained by the following formula: `inv ← (⌹ (⍉coefs)+.×coefs)+.×⍉coefs`.

In this formula, the product `(⍉coefs)+.×coefs` gives a square matrix, even if `coefs` is rectangular.
This is why we can calculate its inverse using the function _matrix inverse_.
If we refer to the definition of _matrix divide_, `(⌹a)+.×b` can be written as `b⌹a`.
So, the formula for `inv` can be rewritten as `inv ← (⍉coefs)⌹(⍉coefs)+.×coefs`.

Let us see what happens if we multiply that formula by `coefs` on the right:

```APL
   inv ≡ (⌹ (⍉coefs)+.×coefs)+.×⍉coefs ←→
←→ inv +.× coefs ≡ (⌹ (⍉coefs)+.×coefs)+.×⍉coefs +.× coefs
```

Now, if we define `U ← (⍉coefs)+.×coefs`, we see that the expression above actually reads

```APL
inv +.× coefs ≡ (⌹U)+.×U
```

`(⌹U)+.×U` is the identity matrix, so we actually see that `inv` is a left inverse of `coefs`, because if we multiply `coefs` with `inv` on the left, we get the identity matrix.

```{admonition} Rule 
:class: tip
The left inverse of a matrix `m` can be computed with `(⌹(⍉m)+.×m)+.×⍉m`.
```

This can be easily verified with the data we used in a previous example:

In [236]:
coefs

25 90 430 2400 1
20 50 87 9000 1
24 12 72 9500 1
28 12 210 4100 1
14 30 144 6500 1
 8 30 91 3300 1
31 120 207 9800 1
17 75 161 4900 1

In [286]:
inv ← (⌹(⍉coefs)+.×coefs)+.×⍉coefs

In [289]:
⌊0.5+inv+.×coefs

1 0 0 0 0
0 1 0 0 0
0 0 1 0 0
0 0 0 1 0
0 0 0 0 1

Once rounded, we can see it is an identity matrix and `inv` really is a left inverse of `coefs`.

(Mathematical-Functions-Pseudo-Right-Inverse)=
#### Pseudo Right Inverse

We just saw that `inv` is a left inverse for `coefs`.
Is it also a right inverse?

We calculated the coefficients in `c` of the least squares line using the formula `c←inv+.×ys`.
Then, we calculated the coordinates of points located on that line with `fits ← coefs+.×c`.
Let us replace `c` by its definition from the first formula, to get `fits ← coefs+.×inv+.×ys`.

If `inv` was a true right inverse of `coefs`, the product `coefs+.×inv` would give the identity matrix and the product `coefs+.×inv+.×ys` would evaluate to `ys`.
However, we know that `ys≢fits` in general, because `coefs+.×inv+.×ys` gives the predicted values on top of the line, whereas `ys` contains the actual values we are trying to fit the line to.
The difference between `fits` and `ys` is such that `+/(ys-fits)*2` is minimised.

In the expression `ys-fits`, let us replace `fits` by `coefs+.×inv+.×ys` and then replace `ys` by `I+.×ys` (`I` is the identity matrix, so this does not change anything) to get `(I+.×ys) - (coefs+.×inv+.×ys)`.
Using `ys` as a common factor, this can be rewritten as `(I-coefs+.×inv)+.×ys`.

If `+/(ys-fits)*2` is minimised, `+/((I-coefs+.×inv)+.×ys)*2` should be minimised too.
This must be true whatever the value of `ys`, including the case where `ys` is 1.
This means that the value of `inv` (calculated independently of `ys`) is such that it minimised the expression `+/(I-coefs+.×inv)*2`.

We can say that `inv` is a pseudo right inverse matrix of `coefs` that minimises `+/(I-coefs+.×inv)*2`.

In APL, _matrix inverse_ has been extended to calculate such a pseudo inverse: `inv ← ⌹coefs`.

(Mathematical-Functions-Summary-of-Left-and-Right-Inverses)=
#### Summary of Left and Right Inverses

In APL, a matrix `M` having more rows than columns has an inverse, which can be obtained by `⌹M`.
This inverse value is strictly equal, within rounding errors, to any of the following formulas:

```APL
      inv ← (⌹(⍉M)+.×M)+.×⍉M
      inv ← (⍉M)⌹(⍉M)+.×M
```

This matrix is a true left inverse of `M`, which means that the product `inv +.× M` gives an identity matrix.

This matrix is also a pseudo right inverse of `M`.
That is to say that the product does not give an identity matrix, but a matrix `J` such that the sum `+/(I-J)*2` is minimised.

the result of this is to minimise any expression of the form `+/((I-M+.×inv)+.×ys)*2`, whatever the value of `ys`.

This last property explains why _domino_ may be used to calculate linear, multi-linear, or polynomial regressions.

(Mathematical-Functions-Scalars-and-Vectors)=
#### Scalars and Vectors

A scalar `s` is treated in such a way that `⌹s` is equal to `÷s` with the exception that `0÷0` is 1 whereas `0⌹0` gives a `DOMAIN ERROR`.

Vectors are treated like single-column matrices:

In [290]:
⎕← cv ← ⌹v ← 2 ¯1 0.5

0.380952 ¯0.190476 0.0952381

In [291]:
cv +.× v

1

In [292]:
v +.× cv

1

(Mathematical-Functions-Total-Array-Ordering)=
### Total Array Ordering

In Dyalog, there is something called _total array ordering_, used by the functions _grade up_ and _down_ and _interval index_.
In the earlier days of Dyalog APL, only simple arrays of the same shape and type could be compared.
Thus, by introducing total array ordering, APL can now order simple and nested arrays of any rank, shape, and type.

Total array ordering is very nuanced. For example, all numbers come before characters:

In [19]:
⍋3 '3'

1 2

In [20]:
⍋234562345243 '3'

1 2

In [21]:
⍋0J1 '3'

1 2

Complex numbers can also be ordered with other numbers:

In [25]:
⍋¯2J¯1 ¯1J¯3 0 0J4 3

1 2 3 4 5

However, mathematics tells us that there is no "natural ordering" for the complex numbers.
This means that we cannot order the complex numbers in a way that preserves the properties of addition and multiplication that we are used to.

For example, a negative number times a positive number yields a negative number.
Recall that "negative" really means "less than zero" and "positive" means "greater than zero".
Below, we can see two complex numbers, one less than zero and the other greater than zero:

In [45]:
⍋¯2J1 0 1J¯2

1 2 3

The product of the two complex numbers shown above is:

In [46]:
1J¯2 × ¯2J1

0J5

And the product is greater than zero:

In [47]:
⍋0 0J5

1 2

We can see that complex numbers are ordered by real part and then by imaginary part:

In [48]:
⍋¯2J¯100 ¯2 ¯2J1 0J¯7654321 0 0J1 3

1 2 3 4 5 6 7

Total array ordering also allows the ordering of nested and mixed arrays:

In [56]:
⍋ ⎕← (⊂⊂⊂0J1 '0J2') 5 ('a',⍳3) ('hey' ('bye' 1 2 (3 'ciao')))

┌───────────────┬─┬───────┬────────────────────────┐
│┌─────────────┐│5│a 1 2 3│┌───┬──────────────────┐│
││┌───────────┐││ │ ││hey│┌───┬─┬─┬────────┐││
│││┌─────────┐│││ │ ││ ││bye│1│2│┌─┬────┐│││
││││┌───┬───┐││││ │ ││ ││ │ │ ││3│ciao││││
│││││0J1│0J2│││││ │ ││ ││ │ │ │└─┴────┘│││
││││└───┴───┘││││ │ ││ │└───┴─┴─┴────────┘││
│││└─────────┘│││ │ │└───┴──────────────────┘│
││└───────────┘││ │ │ │
│└─────────────┘│ │ │ │
└───────────────┴─┴───────┴────────────────────────┘
1 2 3 4

In Dyalog APL, the only simple scalars that are considered by the total array ordering are nulls, numbers, and characters.
In other words, total array ordering does not contemplate namespaces, objects, and object representations.

For example, the following usage of _grade up_ fails because of the namespace in the vector:

In [57]:
⍋1 2 (⎕NS⍬) 4

DOMAIN ERROR
      ⍋1 2(⎕NS ⍬)4
      ∧


(fig-Applied_Maths)=
```{figure} ../res/Applied_Maths.png
---
name: Applied_Maths
---
A man holding a blue hammer chasing a piggy bank.
```


(Mathematical-Functions-Solutions)=
## Solutions

```{solution} ex-encode 
```

We know that `n ⊥ 12 34 60 77 19` is equivalent to `n n n n n ⊥ 12 34 60 77 19`, whatever the value of `n`.
Furthermore, we saw that the weights used in the conversion are given by the expression `⌽1,×\⌽1↓n n n n n`.
So, if `n ← 0`, we get that the weights are

In [237]:
⌽1,×\⌽1↓5⍴0

0 0 0 0 1

This means that the 19 is multiplied by 1 and everything else by 0, so the final result is 19:

In [238]:
0 ⊥ 12 34 60 77 19

19

Therefore, `0⊥vec` gives the last element of a numeric vector.

If `n ← 1`, the weights are

In [239]:
⌽1,×\⌽1↓5⍴1

1 1 1 1 1

This means that every number is multiplied by 1, and thus `1⊥` gives the sum of a vector:

In [240]:
1 ⊥ 12 34 60 77 19

202

In [241]:
+/ 12 34 60 77 19

202

```{solution} ex-hex 
```

Because we know that hexadecimal values will take up 4 characters only, we can safely do `16 16 16 16⊤⍵` to encode a decimal number to base 16:

In [242]:
D2H ← {'0123456789ABCDEF'[1+16 16 16 16⊤⍵]}

In [243]:
(⍪,D2H¨) ¯1+⍳16

┌──┬────┐
│0 │0000│
├──┼────┤
│1 │0001│
├──┼────┤
│2 │0002│
├──┼────┤
│3 │0003│
├──┼────┤
│4 │0004│
├──┼────┤
│5 │0005│
├──┼────┤
│6 │0006│
├──┼────┤
│7 │0007│
├──┼────┤
│8 │0008│
├──┼────┤
│9 │0009│
├──┼────┤
│10│000A│
├──┼────┤
│11│000B│
├──┼────┤
│12│000C│
├──┼────┤
│13│000D│
├──┼────┤
│14│000E│
├──┼────┤
│15│000F│
└──┴────┘

In [244]:
D2H¨ 6748 49679 60281

┌────┬────┬────┐
│1A5C│C20F│EB79│
└────┴────┴────┘

If we did not know how long the hexadecimal number would be, we could use `16⊥⍣¯1⊢` instead:

In [245]:
{'0123456789ABCDEF'[1+16⊥⍣¯1⊢⍵]} 123456789

75BCD15

In [246]:
H2D ← {16⊥¯1+'0123456789ABCDEF'⍳⍵}

In [247]:
H2D¨ '1A5C' 'C20F' 'EB79'

6748 49679 60281

```{solution} ex-sparse-array 
```

Let us work through the solution with the example of the matrix `sa`:

In [248]:
sa ← 5 7⍴0 ⋄ sa[1;2] ← 3 ⋄ sa[4;1] ← 5.6
⎕← sa

0 3 0 0 0 0 0
0 0 0 0 0 0 0
0 0 0 0 0 0 0
5.6 0 0 0 0 0 0
0 0 0 0 0 0 0

To implement the function `Contract`, we need to find the non-zero values of `sa` and their indices:

In [249]:
sa[idx ← ⍸0≠sa]

3 5.6

In [250]:
idx

┌───┬───┐
│1 2│4 1│
└───┴───┘

Now, we need to decode these multidimensional indices into the one-dimensional indices of the ravel, which we can do with the function _decode_ and the shape of the array (`sa`, in our example).
However, indexing is 1-based and _encode_ and _decode_ start counting from 0, so we need to subtract 1 from the indices in the beginning and add 1 at the end:

In [251]:
1+(⍴sa)⊥¯1+⍉↑idx

2 22

We can verify these indices are correct:

In [252]:
(,sa)[2 22]

3 5.6

Now, we can put `Contract` together:

In [253]:
]dinput
Contract ← {
    flat ← 1+(⍴⍵)⊥¯1+⍉↑idx←⍸0≠⍵
    (≢⍴⍵),(⍴⍵),⍵[idx],flat
}

In [254]:
⎕← compressed ← Contract sa

2 5 7 3 5.6 2 22

To implement `Expand`, we work in the other direction.
We start by finding the rank and the shape of the original sparse array:

In [255]:
⎕← rank ← ⊃compressed

2

In [256]:
⎕← shape ← rank↑1↓compressed

5 7

In [257]:
⎕← rest ← compressed↓⍨1+rank

3 5.6 2 22

Now, we need to split the remainder into the values and the indices:

In [258]:
⎕← values ← rest↑⍨0.5×≢rest

3 5.6

In [259]:
⎕← idx ← rest↓⍨0.5×≢rest

2 22

Now, we use _encode_ and the shape of the array to transform the linear indices into the multidimensional ones:

In [260]:
↓⍉1+shape⊤¯1+idx

┌───┬───┐
│1 2│4 1│
└───┴───┘

In [261]:
]dinput
Expand ← {
    shape ← (⊃⍵)↑1↓⍵
    rest ← ⍵↓⍨1+⊃⍵
    values ← rest↑⍨0.5×≢rest
    idx ← rest↓⍨0.5×≢rest
    arr ← shape⍴0
    arr[↓⍉1+shape⊤¯1+idx] ← values
    arr
}

In [262]:
Expand compressed

0 3 0 0 0 0 0
0 0 0 0 0 0 0
0 0 0 0 0 0 0
5.6 0 0 0 0 0 0
0 0 0 0 0 0 0

As another example, here is a cuboid with five non-zero elements:

In [263]:
cuboid ← 3 7 21⍴0
cuboid[(,⍳⍴cuboid)[5?×/⍴cuboid]] ← ⍳5
cuboid

0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 4 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 5 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 3 0 0 0 0 0 0 0

In [264]:
cuboid ≡ Expand Contract cuboid

1

```{solution} ex-random 
```

 - a vector of 12 values between 8 and 30, without duplicates:

In [265]:
7+12?23

21 13 30 29 23 20 28 18 11 25 24 16

 - a 4 by 6 matrix filled with values between 37 and 47, with possible duplicates:

In [266]:
36+?4 6⍴11

39 41 44 38 42 45
47 39 39 40 44 37
44 45 47 43 40 38
43 46 46 47 46 40

- a 5 by 2 matrix filled with values between ¯5 and 5, without duplicates:

In [267]:
¯6+5 2⍴10?11

¯5 ¯3
¯4 0
¯1 4
 2 ¯2
 1 5

```{solution} ex-random-decimals 
```


In [268]:
0.0001×99+?15⍴801

0.0696 0.0865 0.0869 0.0793 0.0118 0.0389 0.0599 0.0475 0.0221 0.0706 0.0165 0.0444 0.0628 0.0359 0.0278

```{solution} ex-read-random 
```

The expression starts with `(10+?10)⍴10` which will create a vector filled with 10, with a random length between 11 and 20.
Then, the final `10+?...` will generate random integers between 1 and 10 and add 10 to those, so the final result is a vector of a random length between 11 and 20 with random elements between 11 and 20.

In [269]:
10+?(10+?10)⍴10

20 11 13 18 16 15 11 14 18 17 15 11

In [270]:
10+?(10+?10)⍴10

15 17 16 12 16 17 11 13 13 20 19 15 16 13 12 18 17

In [271]:
10+?(10+?10)⍴10

18 12 14 15 18 19 18 11 18 16 19 18 19 14 13 15 17 19 12 13

```{solution} ex-random-choose 
```


In [272]:
list ← 12 29 5 44 31 60 8 86

In [273]:
list[5?⍴list]

31 5 86 29 44

```{solution} ex-random-length 
```


In [274]:
2+?(5+?11)⍴38

6 32 18 34 5 34 8 16 9 39 24 27

```{solution} ex-cos-approximation 
```


In [275]:
Cos ← {(⍵*p)-.÷!p←2×0,⍳⍺}

In [276]:
2○○1

¯1

In [277]:
10 Cos ○1

¯1

```{solution} ex-circle 
```

 - `(1○○÷4)*2`:

$\sin \pi/4 = \sqrt{\frac12}$, thus the square of that is 0.5:

In [278]:
(1○○÷4)*2

0.5

 - `2×0.5+¯2○1○0.5`:

`¯2○1○0.5` is $\arccos(\sin \frac12)$, which is $\frac\pi2 - \frac12$, so the whole expression evaluates to $\pi$:

In [279]:
2×0.5+¯2○1○0.5

3.14159

In [280]:
○1

3.14159

```{solution} ex-solve-linear-eqs 
```

We can solve this set of linear equations with the function _matrix divide_.
We just need to be careful when writing the matrix of coefficients.

In [281]:
5 ¯7 2⌹3 3⍴1 ¯1 0 0 1 ¯2 ¯1 0 1

¯2 ¯7 0

```{solution} ex-linear-eqs-compute 
```

To compute the value of $3a + 5b - c$, the easiest thing to do is to start by finding the values of $a$, $b$, and $c$, which we can do by solving the set of linear equations:

In [282]:
⎕← (a b c) ← abc ← 13 ¯6 10⌹3 3⍴1 ¯1 3 ¯2 4 0 1 ¯2 2

2 ¯0.5 3.5

Finally, we can compute the value of that expression in a couple of ways:

In [283]:
(3×a)+(5×b)-c

0

In [284]:
abc+.×3 5 ¯1

0

We could have done this computation directly:

In [285]:
3 5 ¯1+.×13 ¯6 10⌹3 3⍴1 ¯1 3 ¯2 4 0 1 ¯2 2

0